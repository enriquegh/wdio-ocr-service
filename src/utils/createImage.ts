// @ts-ignore
import { encodePNGToStream, decodePNGFromStream, make } from 'pureimage'
import { createReadStream, createWriteStream } from 'fs'
import { Rectangles } from '../types/types'

interface CreateImageData {
  filePath: string;
  height: number;
  lines: any;
  width: number;
}

export default async function createImage(data: CreateImageData) {
  const { filePath, height, lines, width } = data

  const image = await decodePNGFromStream(createReadStream(filePath))
  const canvasImage = make(width, height)
  const context = canvasImage.getContext('2d')
  context.drawImage(
    image,
    // Start at x/y pixels from the left and the top of the image (crop)
    0, 0,
    // 'Get' a (w * h) area from the source image (crop)
    width, height,
    // Place the result at 0, 0 in the canvas,
    0, 0,
    // With as width / height: 100 * 100 (scale)
    width, height
  )

  // Highlight all found texts
  lines.forEach(({ bbox }: { bbox: Rectangles }) => {
    const { right, bottom, left, top } = bbox
    context.beginPath()
    context.fillStyle = 'rgba(57, 170, 86, 0.5)'
    context.fillRect(left, top, right - left, bottom - top)
    context.lineWidth = 2
    context.strokeStyle = '#39aa56'
    context.rect(left, top, right - left, bottom - top)
    context.stroke()
  })

  await encodePNGToStream(canvasImage, createWriteStream(filePath))
}
