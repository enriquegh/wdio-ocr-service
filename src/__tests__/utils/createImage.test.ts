// @ts-ignore
import PureImage from 'pureimage'
import fs from 'fs'
import createImage from '../../utils/createImage'

jest.mock('fs')
jest.mock('pureimage')

interface ContextFunctionArguments {
  drawImageArgs?: string | number[];
  fillRectangleArgs?: string | number[];
  rectangleArgs?: string | number[];
}

const pureImageDecodePNGFromStream = PureImage.decodePNGFromStream
const pureImageEncodePNGToStream = PureImage.encodePNGToStream
const pureImageMake = PureImage.make
let contextFunctionArguments: ContextFunctionArguments[] = [];

describe('createImage', () => {
  beforeEach(() => {
    fs.createReadStream = jest.fn().mockReturnValue('createReadStream')
    fs.createWriteStream = jest.fn().mockReturnValue('createWriteStream')
    pureImageDecodePNGFromStream.mockResolvedValue('pureImageDecodePNGFromStream')
    pureImageEncodePNGToStream.mockResolvedValue('')
    pureImageMake.mockImplementation(() => ({
      getContext: jest.fn().mockImplementation(() => ({
        beginPath: jest.fn().mockImplementation(() => jest.fn()),
        drawImage: jest.fn().mockImplementation(
          (...drawImageArgs) => contextFunctionArguments.push({drawImageArgs})
        ),
        fillRect: jest.fn().mockImplementation(
          (...fillRectangleArgs) => contextFunctionArguments.push({fillRectangleArgs})
        ),
        rect: jest.fn().mockImplementation(
          (...rectangleArgs) => contextFunctionArguments.push({rectangleArgs})
        ),
        stroke: jest.fn().mockImplementation(() => jest.fn()),
      })),
    }))
  })

  afterEach(() => {
    contextFunctionArguments = [];
  })

  it('should call all internal methods with default options', async () => {
    const height = 1792
    const width = 828
    const options = {
      filePath: 'ocr-images/ios-1617168519503.png',
      height,
      width,
    }

    await createImage(options);

    expect(fs.createReadStream).toHaveBeenCalledTimes(1)
    expect(pureImageDecodePNGFromStream).toHaveBeenCalledTimes(1)
    expect(pureImageMake).toHaveBeenCalledTimes(1)
    expect(pureImageMake).toHaveBeenCalledWith(width, height)
    expect(contextFunctionArguments).toMatchSnapshot();
    expect(fs.createWriteStream).toHaveBeenCalledTimes(1)
  })

  it('should be able to crop an image when it is being drawn', async () => {
    const height = 1792
    const width = 828
    const options = {
      filePath: 'ocr-images/ios-1617168519503.png',
      height,
      width,
      top: 300,
      left: 70,
      right: 750,
      bottom: 745
    }

    await createImage(options);

    expect(contextFunctionArguments).toMatchSnapshot();
  })

  it('should be able to crop an image when it is being drawn where top is 0', async () => {
    const height = 1792
    const width = 828
    const options = {
      filePath: 'ocr-images/ios-1617168519503.png',
      height,
      width,
      top: 0,
      left: 0,
      right: 25,
      bottom: 1700
    }

    await createImage(options);

    expect(contextFunctionArguments).toMatchSnapshot();
  })

  it('should be able to draw the lines', async () => {
    const height = 1792
    const width = 828
    const options = {
      filePath: "ocr-images/ios-1617180757754.png",
      height,
      lines: [
        {
          text: 'Username',
          bbox: {
            left: 83,
            top: 326,
            right: 248,
            bottom: 352
          }
        },
        {
          text: 'Password',
          bbox: {
            left: 83,
            top: 454,
            right: 237,
            bottom: 480
          }
        },
        {
          text: 'LOGIN',
          bbox: {
            left: 359,
            top: 669,
            right: 467,
            bottom: 695
          }
        }
      ],
      width,
    }

    await createImage(options);

    expect(contextFunctionArguments).toMatchSnapshot();
  })
})
