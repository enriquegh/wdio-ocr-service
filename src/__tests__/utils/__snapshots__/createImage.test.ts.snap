// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createImage should be able to crop an image when it is being drawn 1`] = `
Array [
  Object {
    "drawImageArgs": Array [
      "pureImageDecodePNGFromStream",
      70,
      300,
      680,
      445,
      70,
      300,
      680,
      445,
    ],
  },
]
`;

exports[`createImage should be able to crop an image when it is being drawn where top is 0 1`] = `
Array [
  Object {
    "drawImageArgs": Array [
      "pureImageDecodePNGFromStream",
      0,
      0,
      25,
      1700,
      0,
      0,
      25,
      1700,
    ],
  },
]
`;

exports[`createImage should be able to draw the lines 1`] = `
Array [
  Object {
    "drawImageArgs": Array [
      "pureImageDecodePNGFromStream",
      0,
      0,
      828,
      1792,
      0,
      0,
      828,
      1792,
    ],
  },
  Object {
    "fillRectangleArgs": Array [
      83,
      326,
      165,
      26,
    ],
  },
  Object {
    "rectangleArgs": Array [
      83,
      326,
      165,
      26,
    ],
  },
  Object {
    "fillRectangleArgs": Array [
      83,
      454,
      154,
      26,
    ],
  },
  Object {
    "rectangleArgs": Array [
      83,
      454,
      154,
      26,
    ],
  },
  Object {
    "fillRectangleArgs": Array [
      359,
      669,
      108,
      26,
    ],
  },
  Object {
    "rectangleArgs": Array [
      359,
      669,
      108,
      26,
    ],
  },
]
`;

exports[`createImage should call all internal methods with default options 1`] = `
Array [
  Object {
    "drawImageArgs": Array [
      "pureImageDecodePNGFromStream",
      0,
      0,
      828,
      1792,
      0,
      0,
      828,
      1792,
    ],
  },
]
`;
